payara/Payara
# APPSERV-47 Adds Custom Watches to Monitoring Console

### Summary
Adds custom watches to client. In contrast to watches collected from `MonitoringWatchSource`s a custom watch is a user defined `Watch` instance. The monitoring client is used to compose such a watch. This can be done in the new page _Watches_ that also lists all watches and allows to edit, delete or disable and enable them.

A disabled watch is simply not actively evaluating data, it is ignored but it will show in the graphs as a outline box instead of a filled one. All watches can be disabled and enabled. This is stored in the added configuration so that the disabled state can be recreated on server startup.

Only custom watches can be delete. Collected (or programmatic) watches cannot be deleted as they originate from the server's code.

Custom watches can be edited. Programmatic watches can be duplicated, which changes the name of the duplicate to _Copy of {Name}_. So this becomes another watch but allows to use the programmatic watch as a template.

To recreate the custom watches on server-startup the watches need to be stored in some form.
I decided to write custom to and from JSON methods for the affected classes. This will allow to put in the proper code that can handle changes to the data structure. The strategy here is to be able to handle missing fields (that would be added) and ignore fields (that would be removed).
The conversion is covered with unit tests.

On server side:
* adds monitoring console configuration to store disabled watches as well as custom watch data
* REST and service API to enable/disable watches
* distinction between custom watches and programmatic watches
* watch states are shared when deriving a watch
* REST and service API to create custom watches
* REST and service API to delete custom watches
* `WatchData` REST API class is not writable to receive custom watch data send by client 

On client side:
* adds a _Watches_ configuration page that lists existing watches, allows to edit, delete or disable/enable them.
* extracts model constants to a separate file for better readability and clarity
* extracts a controller file that encapsulation the client-server communication on client side for better readability and clarity
* fixes/updates formatting of watch conditions (e.g. in alert tables) and watch lists
* uses payara design system colors with better contrast
* fixes alignment of indicator (text message above graph)
* uses configured colours for status messages
* now also triggers status messages from alerts (red=critical, amber=alerting)
* adds new colour defaults for _error_ and _missing_ status
* use of `const` instead of `let` (TLDR; `const` should be default, `let` only when variable is actually changed, `var` is basically never needed or better suited in our "non-legacy" codebase)
* watch indicator at right side of the graph now shows the indicator as outline if the watch is disabled
* `Setting` input with `unit` now can use a function to provide a changing unit.


### `WriteableView` and `ProtectedList`
Tried to improve it to make it more usable. Code also could do with some cleanup.
* added generics
* added missing `@Override` annoations
* removed unnecessary `else` nesting
* made static helpers `static`
* added some `List` methods that would otherwise throw a `UnsupportedOperationException` on the basis of existing method (so no actual transaction management was needed)

As mentioned some of the existing methods of the `ProtectedList` still show some odd behaviour.
When using `set(int, E)` the value at the index would be null next time I loaded the list.

### Tests
Unit tests were added to verify the correctness of the to/from JSON conversion of the `Watch` class and all classes involved.

General information can be taken from the documentation https://github.com/payara/Payara-Server-Documentation/pull/705.

The custom watches feature was tested manually following the steps below:

#### General Setup:
1. build, install and start the server
2. use `set-monitoring-console-configuration --enabled=true` to deploy MC
3. open MC at http://localhost:8080/monitoring-console/
4. make sure browser cache for JS/CSS is cleared for MC's domain
5. check that following pages do exist: _Watches_ (if not most likely a browser cache issue - or get in touch)

#### Testing Disabling Existing Watches
1. Open page _Watches_ in MC
2. click checkbox of one of the listed watches to disable it, e.g. _Metric Collection Duration_
3. for _Metric Collection Duration_ watch open page _Monitoring_ and check the watch indicator on right side of the graph is now drawn as outline instead of filled
4. restart the server, e.g. using asadmin commands
5. refresh page _Watches_ in MC and check disabled watch is still disabled
6. enable the watch again
7. for _Metric Collection Duration_ watch open page _Monitoring_ again and check the indicator is now drawn filled

#### Testing Disabling Custom Watches
1. Open page _Watches_ in MC
2. create a new watch named _Heap Check_ by checking _Unhealthy_, filling in: _If_ `ns:jvm HeapUsage` _in_ `Percent` _is_ `>` _(some value clearly below your current heap usage, see _Core_ page)_, and press _Save or Update_
3. goto _Core_ page and check that the graph shows an alert because heap usage is above the watch level
4. go back to _Watches_ page and disable the created watch by clicking the checkbox on the left or using the cogs menu
5. go back to _Core_ and check that alert is gone (and that indicator is now as outline)
6. go to _Alerts_ page and check the alert is shown in _Past Unacknowledged Alerts_
7. go to _Watches_ page again an enable the watch once more, check on _Core_ page that an alert is triggered again

#### Testing Custom Watches
1. create a custom watch as given in steps for _Testing Disabling Custom Watches_
2. restart the server
3. open _Watches_ page and check the custom watch created is still there

#### Testing Listing of collected Watches
1. open _Watches_ page of MC, with default settings two watches should be listed
2. open admin console and browse to _Configurations_ => _server-config_ => HealthCheck_ tab _CPU Usage_ and check _Enabled_
3. go back to _Watches_ page and check a new watch _CPU Usage_ is listed now

#### Testing new admin command parameters are hidden
run:
```
asadmin> set-monitoring-console-configuration --help
NAME
     set-monitoring-console-configuration

SYNOPSIS
     Usage: set-monitoring-console-configuration [--enabled=true|
     false] 

OPTIONS
     --enabled

Command set-monitoring-console-configuration executed successfully.
```
